Alignment with the newest version of draft-ietf-tls-hybrid-design (#48)
post-quantum-cryptography/draft-kwiatkowski-tls-ecdhe-mlkem: draft-ietf-tls-ecdhe-mlkem.md

# Introduction

The {{hybrid}} document defines a framework for combining traditional key exchanges with next-generation key
exchange in TLS 1.3. The goal of this approach is to provide security against both classical and quantum
adversaries while maintaining compatibility with existing infrastructure and protocols.

This document applies the framework to ML-KEM, a key encapsulation mechanism defined in {{NIST-FIPS-203}},
and specifies code points for the hybrid groups.

## Motivation

This document introduces three new supported groups for hybrid post-quantum key agreements in TLS 1.3: the X25519MLKEM768,
SecP256r1MLKEM768, and SecP384r1MLKEM1024 which combine ML-KEM with ECDH in the manner of {{hybrid}}.

> * The first group uses X25519 {{rfc7748}} and is an update to X25519Kyber768Draft00 {{xyber}}, a widely deployed PQ/T hybrid combiner for TLS v1.3 deployed in 2024.

Author: @bwesterb @csosto-pk I agree this is a case, but does it matter for the draft? 

But I definitely should say that this is the default option. How about something like:

```
The first one uses X25519 {{rfc7748}} and is an update to X25519Kyber768Draft00 {{xyber}}, and will often be the most practical choice when selecting a single option.
```

> * The first group uses X25519 {{rfc7748}} and is an update to X25519Kyber768Draft00 {{xyber}}, a widely deployed PQ/T hybrid combiner for TLS v1.3 deployed in 2024.

Reviewer: Unfortunately, If we're going to have this text, we have to say if it's bitwise identical.

> * The first group uses X25519 {{rfc7748}} and is an update to X25519Kyber768Draft00 {{xyber}}, a widely deployed PQ/T hybrid combiner for TLS v1.3 deployed in 2024.

Author: It is not, bitwise identical. Namely the order of public key and ciphertext is swapped - in the X25519Kyber768Draft00 it is ``X25519||MLKEM`` and in this draft it is ``MLKEM||X25519``.

I would remove ``and is an update to X25519Kyber768Draft00 {{xyber}}`` as those drafts are incompatible anyway (i.e. code points are different).

> * The first group uses X25519 {{rfc7748}} and is an update to X25519Kyber768Draft00 {{xyber}}, a widely deployed PQ/T hybrid combiner for TLS v1.3 deployed in 2024.

Reviewer 2: yes, the whole sentence sounds rather awkward to me.

something like "The first group uses X25519 {{rfc7748}} and is a widely deployed PQ/T hybrid combiner for TLS v1.3." would be sufficient

all-zero shared secret check for X25519, and abort the connection with an
illegal_parameter alert if it fails.

# Discussion

**FIPS-compliance**. All groups permit FIPS-approved key derivation as per {{NIST-SP-800-56C}}
and {{NIST-SP-800-135}}. NIST's special publication 800-56Cr2 {{NIST-SP-800-56C}} approves the
usage of HKDF {{HKDF}} with two distinct shared secrets, with the condition that the first
one is computed by a FIPS-approved key-establishment scheme. FIPS also requires a certified
implementation of the scheme, which will remain more ubiquitous for secp256r1 in the coming years. For
this reason we put the ML-KEM shared secret first in X25519MLKEM768, and the ECDH shared secret
first in SecP256r1MLKEM768 and SecP384r1MLKEM1024.

> **Failures**. We note that ML-KEM algorithm has non-zero probability of encapsulation and decapsulation failures. Those
> failures MAY be handled by clients in a way described in {{Section 4 of hybrid}}.

Reviewer 3: ```suggestion
**Failures**. Any hybrid key establishment failures MAY be handled by clients in a way described in {{Section 4 of hybrid}}.
```
